### PR TITLE
Add convolutional RNN for sentence classification

### PR DESCRIPTION
**Pull request by daemon**

### CODE REVIEW

File: conv_rnn/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+Implementation based on [[1]](http://dl.acm.org/citation.cfm?id=3098140).
+
+### Usage
+
+Run `./getData.sh` to fetch the data. The project structure should now look like this:
+
+```
+├── conv_rnn/
+│   ├── data/
```
Comment (tuzhucheng) [inline conv_rnn/README.md]: Please move all data to https://github.com/castorini/data?

File: conv_rnn/data.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import numpy as np
+import os
```
Comment (tuzhucheng) [inline conv_rnn/data.py]: Let's follow PEP8 standard and move the standard library imports before the third party imports.

File: conv_rnn/getData.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+mkdir data
```
Comment (tuzhucheng) [inline conv_rnn/getData.sh]: Use `mkdir -p` so their isn't an error message if these directories already exist?

File: conv_rnn/train.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    parser.add_argument("--restore", default=False, action="store_true")
+    parser.add_argument("--lr", type=float)
+    parser.add_argument("--weight_decay", type=float)
+    parser.add_argument("--gradient_clip", type=float)
+    parser.add_argument("--seed", type=int)
+    parser.add_argument("--hidden_size", type=int)
+    parser.add_argument("--fc_size", type=int)
+    parser.add_argument("--dropout_prob", type=float)
+    parser.add_argument("--n_feature_maps", type=float)
+    parser.add_argument("--rnn_type", type=str)
```
Comment (tuzhucheng) [inline conv_rnn/train.py]: Use `choices` keyword argument to specify what are the possible choices - see argparse docs for details.

File: conv_rnn/train.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                #scheduler.step(accuracy)
+                if accuracy > conv_rnn.best_dev:
+                    conv_rnn.best_dev = accuracy
+                    torch.save(conv_rnn, kwargs["output_file"])
+                if verbose:
+                    print("Dev set accuracy: {}".format(accuracy))
+                conv_rnn.train()
+    return conv_rnn.best_dev
+
+def do_random_search():
```
Comment (tuzhucheng) [inline conv_rnn/train.py]: Where is this `do_random_search` and ` do_grid_search` called? Please add some driver code that trigger these functions.

File: conv_rnn/train.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+import numpy as np
+import os
+import random 
+import torch
+import torch.nn as nn
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+
+def set_seed(seed=0):
+    np.random.seed(seed)
+    torch.cuda.manual_seed(seed)
```
Comment (tuzhucheng) [inline conv_rnn/train.py]: What happens if code is run in no cuda mode?

File: conv_rnn/test.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    torch.manual_seed(seed)
+    random.seed(seed)
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--input_file", default="saves/model.pt", type=str)
+    parser.add_argument("--data_dir", default="data", type=str)
+    args = parser.parse_args()
+
+    torch.cuda.set_device(1)
```
Comment (tuzhucheng) [inline conv_rnn/test.py]: Expose the gpu number as a parameter?

File: conv_rnn/test.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import argparse
+import data
+import model
+import numpy as np
+import torch
+import torch.nn as nn
+import os
+import random
+
+def set_seed(seed=0):
```
Comment (tuzhucheng) [inline conv_rnn/test.py]: This function also appears in train.py. Let's consolidate it?

File: conv_rnn/test.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--input_file", default="saves/model.pt", type=str)
+    parser.add_argument("--data_dir", default="data", type=str)
+    args = parser.parse_args()
+
+    torch.cuda.set_device(1)
+    set_seed(5)
+    data_loader = data.SSTDataLoader(args.data_dir)
+    conv_rnn = torch.load(args.input_file)
+    conv_rnn.cuda()
```
Comment (tuzhucheng) [inline conv_rnn/test.py]: Should we have a CPU-only option too?

File: conv_rnn/data.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+def sst_tokenize(sentence):
+    extraneous_pattern = re.compile(r"^(--lrb--|--rrb--|``|''|--|\.)$")
+    words = []
+    for word in sentence.split():
+        if re.match(extraneous_pattern, word):
+            continue
+        words.append(word)
+    return words
+
+class SSTDataLoader(object):
```
Comment (tuzhucheng) [inline conv_rnn/data.py]: Please take a look at http://pytorch.org/docs/master/data.html#torch.utils.data.Dataset and http://pytorch.org/docs/master/data.html#torch.utils.data.Dataset. These are PyTorch utility classes that can handle things like data shuffling, batching of different size easily so you don't have to manually deal with them in your code.

File: conv_rnn/model.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            self.bi_rnn = nn.LSTM(embedding_dim, self.hidden_size, 1, batch_first=True, bidirectional=True)
+        elif self.rnn_type == "GRU":
+            self.bi_rnn = nn.GRU(embedding_dim, self.hidden_size, 1, batch_first=True, bidirectional=True)
+        else:
+            raise ValueError
+        self.conv = nn.Conv2d(1, n_fmaps, (1, self.hidden_size * 2))
+        self.dropout = nn.Dropout(dropout)
+        self.fc1 = nn.Linear(n_fmaps + 2 * self.hidden_size, fc_size)
+        self.fc2 = nn.Linear(fc_size, config.get("n_labels", 5))
+        self.epoch = 0
```
Comment (tuzhucheng) [inline conv_rnn/model.py]: Should epoch and best_dev be a part of the model or the training algorithm?

File: conv_rnn/model.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            if not self.no_cuda:
+                h_0 = h_0.cuda()
+                c_0 = c_0.cuda()
+        if self.rnn_type == "LSTM":
+            rnn_seq, rnn_out = self.bi_rnn(x, (h_0, c_0)) # shape: (batch, seq len, 2 * hidden_size), (2, batch, hidden_size)
+            rnn_out = rnn_out[0] # (h_0, c_0)
+        else:
+            rnn_seq, rnn_out = self.bi_rnn(x, h_0) # shape: (batch, 2, hidden_size)
+        rnn_out.data = rnn_out.data.permute(1, 0, 2)
+        x = self.conv(rnn_seq.unsqueeze(1)).squeeze(3) # shape: (batch, channels, seq len)
```
Comment (tuzhucheng) [inline conv_rnn/model.py]: Nice comments of dimensions at end!

File: conv_rnn/model.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        x = nn_func.relu(x) # shape: (batch, channels, seq len)
+        x = nn_func.max_pool1d(x, x.size(2)) # shape: (batch, channels)
+        out = [t.squeeze(1) for t in rnn_out.chunk(2, 1)]
+        out.append(x)
+        x = torch.cat(out, 1).squeeze(2)
+        #x = self.dropout(x)
```
Comment (tuzhucheng) [inline conv_rnn/model.py]: Let's avoid having commented out code

File: conv_rnn/train.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    parser.add_argument("--mbatch_size", type=int)
+    parser.add_argument("--n_epochs", type=int)
+    parser.add_argument("--restore", default=False, action="store_true")
+    parser.add_argument("--lr", type=float)
+    parser.add_argument("--weight_decay", type=float)
+    parser.add_argument("--gradient_clip", type=float)
+    parser.add_argument("--seed", type=int)
+    parser.add_argument("--hidden_size", type=int)
+    parser.add_argument("--fc_size", type=int)
+    parser.add_argument("--dropout_prob", type=float)
```
Comment (tuzhucheng) [inline conv_rnn/train.py]: Would it be better to move all the default values here (specify using `default` kwarg) instead of elsewhere (i.e. at the top of `train`?)

File: conv_rnn/README.md (hunk 1 above)
Comment (daemon) [reply conv_rnn/README.md]: Oh I used the script from https://github.com/castorini/Castor/tree/master/kim_cnn

**Comment by daemon** [comment]
Very detailed :+1:

**Comment by lintool** [comment]
I expect nothing less!

**Comment by daemon** [comment]
I've made all the relevant changes.

**Comment by daemon** [comment]
Code now uses Dataset and DataLoader. An observation first though:

I noticed that I hadn't been zeroing the gradients each time before `optimizer.step()`, so I corrected it. However, the model does quite a bit worse with the fix--I tried vanilla hyperparameters too. For now, I've reverted to the incorrect zero_grad() placed in the epoch loop. 

I'll look into this later.
